mip/driver_imx_rt1020.c

```diff
@@ -0,0 +1,302 @@
+                  | (uint32_t)(0x2 << 16) \
+                  | (uint32_t)(val);
+  ENET->MMFR = phy_transaction;
+  wait_phy_complete();
+}
+
+// FEC RX/TX descriptors (Enhanced descriptor not enabled)
+// Descriptor buffer structure, little endian
+
+typedef struct enet_bd_struct_def
+{
+    uint16_t length;  // Data length
+    uint16_t control; // Control and status
+    uint32_t *buffer; // Data ptr
+} enet_bd_struct_t;
+
+// Descriptor and buffer globals, in non-cached area, 64 bits aligned.
+
+__attribute__((section("NonCacheable,\"aw\",%nobits @"))) enet_bd_struct_t rx_buffer_descriptor[(ENET_RXBD_NUM)] __attribute__((aligned((64U))));
+__attribute__((section("NonCacheable,\"aw\",%nobits @"))) enet_bd_struct_t tx_buffer_descriptor[(ENET_TXBD_NUM)] __attribute__((aligned((64U))));
+
+uint8_t rx_data_buffer[(ENET_RXBD_NUM)][((unsigned int)(((ENET_RXBUFF_SIZE)) + (((64U))-1U)) & (unsigned int)(~(unsigned int)(((64U))-1U)))] __attribute__((aligned((64U))));
+uint8_t tx_data_buffer[(ENET_TXBD_NUM)][((unsigned int)(((ENET_TXBUFF_SIZE)) + (((64U))-1U)) & (unsigned int)(~(unsigned int)(((64U))-1U)))] __attribute__((aligned((64U))));
+
+// Initialise driver imx_rt1020
+
+static bool mip_driver_imx_rt1020_init(uint8_t *mac, void *data) {
+  // ENET Reset, wait complete
+  ENET->ECR |= BIT_SET(0);
+  while((ENET->ECR & BIT_SET(0)) != 0) {}
+
+  // Re-latches the pin strapping pin values
+  ENET->ECR |= BIT_SET(0);
+  while((ENET->ECR & BIT_SET(0)) != 0) {}
+
+  // Setup MII/RMII MDC clock divider (<= 2.5MHz).
+  ENET->MSCR = 0x130; // HOLDTIME 2 clk, Preamble enable, MDC MII_Speed Div 0x30
+  eth_write_phy(PHY_ADDR, PHY_BCR, 0x8000); // PHY W @0x00 D=0x8000 Soft reset
+  while (eth_read_phy(PHY_ADDR, PHY_BSR) & BIT_SET(15)) {delay(0x5000);} // Wait finished poll 10ms
+
+  // PHY: Start Link
+  {
+    eth_write_phy(PHY_ADDR, PHY_BCR, 0x1200); // PHY W @0x00 D=0x1200 Autonego enable + start
+    eth_write_phy(PHY_ADDR, 0x1f, 0x8180);    // PHY W @0x1f D=0x8180 Ref clock 50 MHz at XI input
+
+    // Auto configuration
+    {
+      uint32_t linkup = 0;
+      int linkup_tentatives = 5;
+      while (!linkup && linkup_tentatives-- > 0) {
+        linkup = mip_driver_imx_rt1020_up(NULL);
+        delay(0x500000); // Approx 1s
+      }
+
+      if (!linkup) {
+        MG_ERROR(("Error: Link didn't come up"));
+      }
+      else {
+        MG_INFO(("Link up"));
+      }
+    }
+
+    // PHY MII configuration
+    {
+      uint32_t bcr = eth_read_phy(PHY_ADDR, PHY_BCR);
+      bcr &= BIT_CLEAR(10); // Isolation -> Normal
+      eth_write_phy(PHY_ADDR, PHY_BCR, bcr);
+
+      uint32_t linkup = 0;
+      int linkup_tentatives = 5;
+      while (!linkup && linkup_tentatives-- > 0) {
+        linkup = mip_driver_imx_rt1020_up(NULL);
+        delay(0x500000); // Approx 1s
+      }
+
+      if (!linkup) {
+        MG_ERROR(("Error: Link not ready"));
+      }
+      else {
+        MG_INFO(("Link ready"));
+      }
+    }
+  }
+
```

Comment:
Is this really necessary ? We poll the PHY every second for link status

Reply:
I see you suggest, so let just DHCP/Network code in general do the timout/reties and delete this graceful retry in init right ?
You sure that's better ? i can update of course.

Reply:
Ok, i updated it